src/backends/perf_pt/mod.rs

```diff
@@ -323,65 +321,92 @@ impl Drop for PerfPTTrace {
     }
 }
 
-/// Configures a [`PerfPTTracer`](struct.PerfPTTracer.html).
+/// Configures a [`PerfPTThreadTracer`](struct.PerfPTThreadTracer.html).
 ///
 // Must stay in sync with the C code.
+#[derive(Clone)]
 #[repr(C)]
 pub struct PerfPTConf {
-    // Thread ID to trace.
-    target_tid: pid_t,
     // Data buffer size, in pages. Must be a power of 2.
-    data_bufsize: size_t,
+    pub data_bufsize: size_t,
     // AUX buffer size, in pages. Must be a power of 2.
-    aux_bufsize: size_t,
+    pub aux_bufsize: size_t,
     // The initial trace buffer size (in bytes) for new traces.
-    new_trace_bufsize: size_t,
+    pub new_trace_bufsize: size_t,
 }
 
-impl PerfPTConf {
-    /// Creates a new configuration with defaults.
-    pub fn new() -> Self {
+impl Default for PerfPTConf {
+    fn default() -> Self {
         Self {
-            target_tid: linux_gettid(),
             data_bufsize: 64,
             aux_bufsize: 1024,
             new_trace_bufsize: 1024 * 1024, // 1MiB
         }
     }
+}
 
-    /// Select which thread to trace.
-    ///
-    /// By default, the current thread is traced.
-    ///
-    /// The `tid` argument is a Linux thread ID. Note that Linux re-uses the `pid_t` type, but that
-    /// PIDs are distinct from TIDs.
-    pub fn target_tid(mut self, pid: pid_t) -> Self {
-        self.target_tid = pid;
-        self
-    }
+pub struct PerfPTTracer {
+    config: PerfPTConf,
+}
+
+impl PerfPTTracer {
+    fn check_perf_perms() -> Result<(), HWTracerError> {
+        if unsafe { geteuid() } == 0 {
+            // Root can always trace.
+            return Ok(());
+        }
+
+        let mut f = File::open(&PERF_PERMS_PATH)?;
+        let mut buf = String::new();
+        f.read_to_string(&mut buf)?;
+        let perm = buf.trim().parse::<i8>()?;
+        if perm != -1 {
+            let msg = format!("Tracing not permitted: you must be root or {} must contain -1",
+                           PERF_PERMS_PATH);
+            return Err(HWTracerError::Permissions(msg));
+        }
 
-    /// Set the PT data buffer size (in pages).
-    pub fn data_bufsize(mut self, size: usize) -> Self {
-        self.data_bufsize = size as size_t;
-        self
+        Ok(())
     }
+}
+
+impl Tracer for PerfPTTracer {
+    fn from_backend_config(config: &BackendConfig) -> Result<Box<Tracer>, HWTracerError> where Self: Sized {
+        let mut ppt_config = PerfPTConf::default();
+
+        if config.perf_data_bufsize.is_some() {
+            ppt_config.data_bufsize = config.perf_data_bufsize.unwrap();
+        }
+        if config.perf_aux_bufsize.is_some() {
+            ppt_config.aux_bufsize = config.perf_aux_bufsize.unwrap();
+        }
+        if config.new_trace_bufsize.is_some() {
+            ppt_config.new_trace_bufsize = config.new_trace_bufsize.unwrap();
+        }
+
+        // Check for inavlid configuration.
+        fn power_of_2(v: size_t) -> bool {
```

Comment:
Isn't `size_t` unsigned in Rust? If so the `v <= 0` check is unneeded.

Reply:
Well I never!

```
type size_t = usize;
```

But that doesn't match C! And this is in the libc crate. Suspicious...

Reply:
Ah no wait, in C you have `ssize_t`. Yes, you are right.